src/js/components/FileInput/FileInput.js

```diff
@@ -221,198 +295,165 @@ const FileInput = forwardRef(
-          ref={inputRef}
-          type="file"
-          id={id}
-          name={name}
-          multiple={multiple}
-          disabled={disabled}
-          plain
-          rightOffset={rightOffset}
-          {...rest}
-          onDragOver={() => setDragOver(true)}
-          onDragLeave={() => setDragOver(false)}
-          onChange={(event) => {
-            event.persist();
-            const fileList = event.target.files;
-            const nextFiles = multiple ? [...files] : [];
-            for (let i = 0; i < fileList.length; i += 1) {
-              // avoid duplicates
-              const existing =
-                nextFiles.filter(
-                  (file) =>
-                    file.name === fileList[i].name &&
-                    file.size === fileList[i].size,
-                ).length > 0;
-              if (!existing) nextFiles.push(fileList[i]);
-            }
-            setFiles(nextFiles);
-            setDragOver(false);
-            if (onChange) onChange(event, { files: nextFiles });
-          }}
-        />
-      </ContentsBox>
+            ))}
+          <StyledFileInput
+            ref={inputRef}
+            type="file"
+            id={id}
+            name={name}
+            multiple={multiple}
+            disabled={disabled}
+            plain
+            rightOffset={rightOffset}
+            {...rest}
+            onDragOver={() => setDragOver(true)}
+            onDragLeave={() => setDragOver(false)}
+            onChange={(event) => {
+              event.persist();
+              const fileList = event.target.files;
+              let filesToAdd;
+              if (!files.length) {
+                if (multiple && max && fileList.length >= max) {
+                  const newFileList = [];
+                  for (let i = 0; i < max; i += 1) {
+                    newFileList.push(fileList[i]);
+                  }
+                  filesToAdd = newFileList;
+                } else {
+                  filesToAdd = fileList;
+                }
+              } else if (
+                multiple &&
+                max &&
+                fileList.length >= max - files.length
+              ) {
+                const newFileList = [];
+                for (let i = 0; i < max - files.length; i += 1) {
+                  newFileList.push(fileList[i]);
+                }
+                filesToAdd = newFileList;
+              } else {
+                filesToAdd = fileList;
+              }
+              const nextFiles = multiple ? [...files] : [];
+              for (let i = 0; i < filesToAdd.length; i += 1) {
+                // avoid duplicates
+                const existing =
+                  nextFiles.filter(
+                    (file) =>
+                      file.name === fileList[i].name &&
+                      file.size === fileList[i].size,
+                  ).length > 0;
+                if (!existing) nextFiles.push(filesToAdd[i]);
+              }
```

Comment:
This code seems to be trying to avoid making nextFiles larger than max. I think it has an interesting corner-case of when one of the new files is a duplicate of one of the existing files and a max is defined. How about instead just slicing the result to keep it under max. For example:
```jsx
let nextFiles = multiple ? [...files] : [];
for (let i = 0; i < fileList.length; i += 1) {
  // avoid duplicates
  const existing =
  nextFiles.filter(
    (file) =>
      file.name === fileList[i].name &&
      file.size === fileList[i].size,
  ).length > 0;
  if (!existing) nextFiles.push(fileList[i]);
}
if (nextFiles.length > max) {
  nextFiles = nextFiles.slice(0, max);
}
```
This also assumes max was defaulted to `Number.POSITIVE_INFINITY`.

Reply:
I will refactor this one. Thank you.